### ISSUE TITLE
float8 Grouped MM kernels

### ISSUE THREAD

**Issue by BIGBALLON** [pytorch/torchtitan#1716]
**Is there any plan to support float8 Grouped MM for llama4 / qwen3 MoE model training?**

Currently, the available Grouped GEMM kernels only support float16, and they do not work with float8.

For training large MoE models such as LLaMA4 and Qwen3, float8 is increasingly important to reduce memory footprint and improve training efficiency. However, the lack of float8 Grouped GEMM support becomes a bottleneck when scaling up MoE training.

**Feature request:**

Add support for float8 Grouped GEMM kernels.

Ensure compatibility with MoE training workloads (e.g., LLaMA4, Qwen3).

This would enable more efficient large-scale MoE training under float8 precision.

**Comment by tianyu-l** [comment]
@danielvegamyhre

**Comment by danielvegamyhre** [comment]
@BIGBALLON we have support for (1) float8 rowwise dynamic quantization + grouped gemm and (2) mxfp8 dynamic quantization + grouped gemm (both differentiable) via [torchao _scaled_grouped_mm](https://github.com/pytorch/ao/blob/c801f10330a1b1261687b702fdd055d033f230d1/torchao/prototype/moe_training/scaled_grouped_mm.py#L40).  Under the hood, this does dynamic quantization on the inputs then uses `torch._scaled_grouped_mm` to dispatch to the appropriate CUTLASS kernel for the fp8/mxfp8 grouped gemm.

torchtitan integration is WIP. Let me know if you have any follow up questions.

Please note low precision MoE training is still in progress, and while we have promising preliminary results (measured mxfp8 at ~1.27x speedup over bf16 in single node, FSDP only training of 2 layer Llama4 model with seq_len=8192), it is still in prototype phase and not officially supported yet. 

Please feel free to reach out with any questions or feedback!

**Comment by BIGBALLON** [comment]
@danielvegamyhre 

Thanks a lot for your quick reply — I’m really looking forward to seeing the full integration on Titan.

I have one more question: what’s the difference between rowwise and blockwise quantization? I noticed the integration here uses rowwise — is there any reference or material you’d recommend to learn more about this?

Will there be future updates to the documentation or examples that could help guide us in using Titan for low-precision MoE training more effectively?

Thanks again for sharing the details — this is really helpful!

**Comment by danielvegamyhre** [comment]
> I have one more question: what’s the difference between rowwise and blockwise quantization? I noticed the integration here uses rowwise — is there any reference or material you’d recommend to learn more about this?

- Rowwise = scaling group spans the entire axis, i.e., (1 x num_cols).
- Blockwise = deepseekv3 style scaling, where input activations are scaled with (1 x 128 granularity) and weights scaled with (128x128 granularity).

> Will there be future updates to the documentation or examples that could help guide us in using Titan for low-precision MoE training more effectively?

Yes, definitely! We are wrapping up some final changes, fixes, and planning some scale testing over the next few weeks. Once things are finalized we'll add documentation and benchmarks for your reference. I'll make a note to update this issue when we have an update on this.

**Comment by BIGBALLON** [comment]
@danielvegamyhre Thank you very much for the guidance. My immediate goal is to train MoE models such as Llama-4 or Qwen3-MoE on H-series GPUs with Titan, in FP8 precision.

If feasible, I’d like to start with tensor-wise FP8 training for MoE and then move on to the block-wise kernel once it’s ready.

If your schedule is too tight for the block-wise route, would it be possible to land tensor-wise MoE support first?

I’m also happy to contribute directly, but I’m not sure where to begin. If you could point me to the right files/design decisions, I’d be glad to help implement or test whatever is needed.

**Comment by danielvegamyhre** [comment]
> [@danielvegamyhre](https://github.com/danielvegamyhre) Thank you very much for the guidance. My immediate goal is to train MoE models such as Llama-4 or Qwen3-MoE on H-series GPUs with Titan, in FP8 precision.
> 
> If feasible, I’d like to start with tensor-wise FP8 training for MoE and then move on to the block-wise kernel once it’s ready.
> 
> If your schedule is too tight for the block-wise route, would it be possible to land tensor-wise MoE support first?
> 
> I’m also happy to contribute directly, but I’m not sure where to begin. If you could point me to the right files/design decisions, I’d be glad to help implement or test whatever is needed.

So we only have fp8 scaled grouped gemm support for rowwise scaling, not tensorwise. You can try MoE fp8 rowwise training with torchtitan right now, if you want. Note you'll need to enable torch.compile for good performance. Here is an example command:

Example:

```
 NGPU=4 CONFIG_FILE="./torchtitan/experiments/llama4/train_configs/debug_model.toml" ./run_train.sh --training.steps=50 --model.converters="float8" --float8.recipe_name="rowwise" --mx.moe_fqns_prototype="experts" --compile.enable
```

**Comment by BIGBALLON** [comment]
> > [@danielvegamyhre](https://github.com/danielvegamyhre) Thank you very much for the guidance. My immediate goal is to train MoE models such as Llama-4 or Qwen3-MoE on H-series GPUs with Titan, in FP8 precision.
> > If feasible, I’d like to start with tensor-wise FP8 training for MoE and then move on to the block-wise kernel once it’s ready.
> > If your schedule is too tight for the block-wise route, would it be possible to land tensor-wise MoE support first?
> > I’m also happy to contribute directly, but I’m not sure where to begin. If you could point me to the right files/design decisions, I’d be glad to help implement or test whatever is needed.
> 
> So we only have fp8 scaled grouped gemm support for rowwise scaling, not tensorwise. You can try MoE fp8 rowwise training with torchtitan right now, if you want. Note you'll need to enable torch.compile for good performance. Here is an example command:
> 
> Example:
> 
> ```
>  NGPU=4 CONFIG_FILE="./torchtitan/experiments/llama4/train_configs/debug_model.toml" ./run_train.sh --training.steps=50 --model.converters="float8" --float8.recipe_name="rowwise" --mx.moe_fqns_prototype="experts" --compile.enable
> ```

@danielvegamyhre It seems that when I train Llama-4 with FP8, the speed is actually slower than with BF16. Is this expected, or could there be something misconfigured on my side?

**Comment by danielvegamyhre** [comment]
> > > [@danielvegamyhre](https://github.com/danielvegamyhre) Thank you very much for the guidance. My immediate goal is to train MoE models such as Llama-4 or Qwen3-MoE on H-series GPUs with Titan, in FP8 precision.
> > > If feasible, I’d like to start with tensor-wise FP8 training for MoE and then move on to the block-wise kernel once it’s ready.
> > > If your schedule is too tight for the block-wise route, would it be possible to land tensor-wise MoE support first?
> > > I’m also happy to contribute directly, but I’m not sure where to begin. If you could point me to the right files/design decisions, I’d be glad to help implement or test whatever is needed.
> > 
> > 
> > So we only have fp8 scaled grouped gemm support for rowwise scaling, not tensorwise. You can try MoE fp8 rowwise training with torchtitan right now, if you want. Note you'll need to enable torch.compile for good performance. Here is an example command:
> > Example:
> > ```
> >  NGPU=4 CONFIG_FILE="./torchtitan/experiments/llama4/train_configs/debug_model.toml" ./run_train.sh --training.steps=50 --model.converters="float8" --float8.recipe_name="rowwise" --mx.moe_fqns_prototype="experts" --compile.enable
> > ```
> 
> [@danielvegamyhre](https://github.com/danielvegamyhre) It seems that when I train Llama-4 with FP8, the speed is actually slower than with BF16. Is this expected, or could there be something misconfigured on my side?

Careful using the debug model, the dims are extremely tiny and unrealistic (dim=256).

Use a standard llama4 17bx16e config.

Even then, fp8 rowwise is only faster when experts per device is low (e.g., 1) due to this issue: https://github.com/pytorch/ao/issues/2880 

The best speedup I've measured for fp8 rowwise was only ~10%, for single node FSDP-only training with experts=[1 or 2] and dim=5120 (full dim size for llama4). (Artifically reduced number of layers to 2, to avoid OOMs).


Mxfp8 results are much more promising, but it's not ready yet. Will share an update when we have one in the coming 2-3 weeks.

**Comment by BIGBALLON** [comment]
@danielvegamyhre Is mxfp8 limited to B-series GPUs? I only have H-series cards, but I’m really hoping Titan will support MoE FP8 training on H-series GPUs as well. Thank you very much!

**Comment by danielvegamyhre** [comment]
> [@danielvegamyhre](https://github.com/danielvegamyhre) Is mxfp8 limited to B-series GPUs? I only have H-series cards, but I’m really hoping Titan will support MoE FP8 training on H-series GPUs as well. Thank you very much!

We will! Can you share your LLama4 model config/toml and launch command you used that yielded poor performance?

**Comment by BIGBALLON** [comment]
Thank you for the update! @danielvegamyhre 

I still have a couple of questions:

``` python
    h = F.silu(
        torch._grouped_mm(x.bfloat16(), w1.bfloat16().transpose(-2, -1), offs=offsets)
    )
    h = h * torch._grouped_mm(
        x.bfloat16(), w3.bfloat16().transpose(-2, -1), offs=offsets
    )
    out = torch._grouped_mm(h, w2.bfloat16().transpose(-2, -1), offs=offsets).type_as(x)

```
 
- Why is it necessary to convert the inputs to bfloat16() in this context?
- How does the hook for [torchao _scaled_grouped_mm](https://github.com/pytorch/ao/blob/c801f10330a1b1261687b702fdd055d033f230d1/torchao/prototype/moe_training/scaled_grouped_mm.py#L40)
 work, and how can we verify that it's functioning as expected?

Looking forward to your insights!
